# 복습 - 답

In [1]:
import pandas as pd

In [2]:
from pandas import Series, DataFrame

In [3]:
import numpy as np
np.random.seed(12345)
import matplotlib.pyplot as plt
plt.rc('figure', figsize=(10, 6))
PREVIOUS_MAX_ROWS = pd.options.display.max_rows
pd.options.display.max_rows = 20
np.set_printoptions(precision=4, suppress=True)

### 1. 데이터 로딩: pd.read_csv()

 'datasets/exercise/insurance.csv'파일 읽어서 df에 할당하기

In [4]:
df = pd.read_csv('datasets/exercise/insurance.csv')

### 2. df의 상위 5개 항목 출력

In [5]:
df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

### 3. df의 요약(평균, 4분위수, 전체 갯수 등)

In [54]:
df.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

### 4. 신규 컬럼(double bmi)을 생성하고, bmi의 2배가 되는 값을 할당하기

In [55]:
df['double bmi'] = df['bmi'] * 2
df.head()

age     sex     bmi  children smoker     region      charges  double bmi
0   19  female  27.900         0    yes  southwest  16884.92400       55.80
1   18    male  33.770         1     no  southeast   1725.55230       67.54
2   28    male  33.000         3     no  southeast   4449.46200       66.00
3   33    male  22.705         0     no  northwest  21984.47061       45.41
4   32    male  28.880         0     no  northwest   3866.85520       57.76

### 5. 신규 컬럼(debt)을 생성하고, 0으로 채우기

In [56]:
df['debt'] = 0
df.head()

age     sex     bmi  children smoker     region      charges  double bmi  \
0   19  female  27.900         0    yes  southwest  16884.92400       55.80   
1   18    male  33.770         1     no  southeast   1725.55230       67.54   
2   28    male  33.000         3     no  southeast   4449.46200       66.00   
3   33    male  22.705         0     no  northwest  21984.47061       45.41   
4   32    male  28.880         0     no  northwest   3866.85520       57.76   

   debt  
0     0  
1     0  
2     0  
3     0  
4     0

### 6. debt에 0부터 값이 증가하도록 채우기

In [58]:
count = len(df)   #1338

In [60]:
A = [] # 빈 리스트를 만들어라..

# 빈 리스트에 넣어야 하는 규칙을 만들어라.
for i in range(count):
    
    A.append(i)

A

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


In [61]:
A = np.array([A]) # 각 행들끼리 구분이 안되는 1D 차원이다.
A.shape # array는 transpose가 안됀다. 왜냐하면 '행렬'개념이 아니기 때문이다.

(1, 1338)

In [62]:
A = np.matrix(A)  # 행렬 기준으로 바꾸려면은 matrix를 써야한다.
A

# ex) [1,2] 는 1X2 행렬이다. 따라서 [1,2,3,4,5,,,,,,,,,,1338]은 1X1338 행렬인 것이다.
# array(1,2) = [1,2]      는 (2,)이 나오는데, 데이터가 2개가 있다는 말이다.. (1과 2... so, 2개의 데이터)
# array([1,2]) = [1,2]    는 (1,2)이 나오는데 [1,2]가 1개이고 그안에 데이터가 2개가 들어갔다. 즉 1X2 행렬이라는 말이다.  이는 matrix(1,2)과 같다.

# matrix(1,2) = [1,2]  라는 뜻! 1X2 행렬이라는 말

# matrix([[   0,    1,    2, ..., 1335, 1336, 1337]])     ===> 1X1338 행렬이다.. 

matrix([[   0,    1,    2, ..., 1335, 1336, 1337]])

In [63]:
#a=np.array([[1,2],[2,3]])
#a.shape
a=np.array([[1,2,3],[2,3,3]])
a.shape     ### =====> 2개의 데이터가 있고, 각각 3개의 데이터를 가지고 있다.

(2, 3)

In [64]:
A = np.transpose(A)  # 1X1338 행렬을 1338X1 행렬로 변환하라
A.shape # matrix는 '행렬'개념이므로, transpose가 가능하다.

(1338, 1)

In [65]:
debt = A
df['debt'] = debt
df.head()

age     sex     bmi  children smoker     region      charges  double bmi  \
0   19  female  27.900         0    yes  southwest  16884.92400       55.80   
1   18    male  33.770         1     no  southeast   1725.55230       67.54   
2   28    male  33.000         3     no  southeast   4449.46200       66.00   
3   33    male  22.705         0     no  northwest  21984.47061       45.41   
4   32    male  28.880         0     no  northwest   3866.85520       57.76   

   debt  
0     0  
1     1  
2     2  
3     3  
4     4

In [66]:
#df['debt'] = np.arange(len(df),0,-1)
df['debt'] = np.arange(0, len(df))
df

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

      double bmi  debt  
0          55.80     0  
1          67.54     1  
2          66.00     2  
3          45.41     3  
4          57.76     4  
...          ...   ...  
1333       61.94  1333  
1334       63.84  1334  
1335       73.70  1335  
1336       51.60  1336  
1337       58.14  1337  

[1338 rows x 9 columns]

### 7. 1번 인덱스의 정보 출력:iloc

In [10]:
# 특정 row에 해당하는 값 : loc 
df.iloc[1]

age                  18
sex                male
bmi               33.77
children              1
smoker               no
region        southeast
charges         1725.55
double bmi        67.54
debt                  1
Name: 1, dtype: object

### 8. 3,5,8번 인덱스의 'debt'값을 12, 13, 14로 설정하기

In [11]:
val = pd.Series([12, 13, 14], index=[3, 5, 8])
df['debt'] = val
df

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

      double bmi  debt  
0          55.80   NaN  
1          67.54   NaN  
2          66.00   NaN  
3          45.41  12.0  
4          57.76   NaN  
...          ...   ...  
1333       61.94   NaN  
1334       63.84   NaN  
1335       73.70   NaN  
1336       51.60   NaN  
1337       58.14   NaN  

[1338 rows x 9 columns]

### 9. 'double bmi'열 삭제하기

In [12]:
del df['double bmi']
# df.drop(['double bmi'], axis=1)

### 10. 'age'열, 'sex'열만 출력하기

In [13]:
F

age     sex
0      19  female
1      18    male
2      28    male
3      33    male
4      32    male
...   ...     ...
1333   50    male
1334   18  female
1335   18  female
1336   21  female
1337   61  female

[1338 rows x 2 columns]

### 11. 함수 생성, 적용(apply/map)

나이를 1살씩 증가시키기

In [14]:
f = lambda x: x+1
df['age'].apply(f)

0       20
1       19
2       29
3       34
4       33
        ..
1333    51
1334    19
1335    19
1336    22
1337    62
Name: age, Length: 1338, dtype: int64

### 12. 함수 생성, 적용(apply/map)

charges의 표시형식을 소수점 아래 2자리로 바꾸기

In [15]:
format = lambda x: '%.2f' % x
df['charges'].map(format)

0       16884.92
1        1725.55
2        4449.46
3       21984.47
4        3866.86
          ...   
1333    10600.55
1334     2205.98
1335     1629.83
1336     2007.94
1337    29141.36
Name: charges, Length: 1338, dtype: object

### 13. 함수 생성, 적용(applymap)

데이터프레임 전체 값을 대괄호[ ] 로 감싸기

예: 	[19]	[female]	[27.9]	[0]	[yes]	[southwest]	[16884.924]

In [16]:
ff = lambda x: '[' + str(x) + ']'
df.applymap(ff)

age       sex       bmi children smoker       region        charges  \
0     [19]  [female]    [27.9]      [0]  [yes]  [southwest]    [16884.924]   
1     [18]    [male]   [33.77]      [1]   [no]  [southeast]    [1725.5523]   
2     [28]    [male]    [33.0]      [3]   [no]  [southeast]     [4449.462]   
3     [33]    [male]  [22.705]      [0]   [no]  [northwest]  [21984.47061]   
4     [32]    [male]   [28.88]      [0]   [no]  [northwest]    [3866.8552]   
...    ...       ...       ...      ...    ...          ...            ...   
1333  [50]    [male]   [30.97]      [3]   [no]  [northwest]   [10600.5483]   
1334  [18]  [female]   [31.92]      [0]   [no]  [northeast]    [2205.9808]   
1335  [18]  [female]   [36.85]      [0]   [no]  [southeast]    [1629.8335]   
1336  [21]  [female]    [25.8]      [0]   [no]  [southwest]     [2007.945]   
1337  [61]  [female]   [29.07]      [0]  [yes]  [northwest]   [29141.3603]   

        debt  
0      [nan]  
1      [nan]  
2      [nan]  
3     [12.0]  
4      [nan]  
...      ...  
1333   [nan]  
1334   [nan]  
1335   [nan]  
1336   [nan]  
1337   [nan]  

[1338 rows x 8 columns]

### 14. 컬럼 정렬(알파벳 순서대로) : sort_index
age, bmi, charges... 순으로 정렬

In [17]:
df.sort_index(axis=1)

age     bmi      charges  children  debt     region     sex smoker
0      19  27.900  16884.92400         0   NaN  southwest  female    yes
1      18  33.770   1725.55230         1   NaN  southeast    male     no
2      28  33.000   4449.46200         3   NaN  southeast    male     no
3      33  22.705  21984.47061         0  12.0  northwest    male     no
4      32  28.880   3866.85520         0   NaN  northwest    male     no
...   ...     ...          ...       ...   ...        ...     ...    ...
1333   50  30.970  10600.54830         3   NaN  northwest    male     no
1334   18  31.920   2205.98080         0   NaN  northeast  female     no
1335   18  36.850   1629.83350         0   NaN  southeast  female     no
1336   21  25.800   2007.94500         0   NaN  southwest  female     no
1337   61  29.070  29141.36030         0   NaN  northwest  female    yes

[1338 rows x 8 columns]

### 15. 값 정렬(age) : sort_values
나이 순으로 정렬

In [18]:
df.sort_values(by='age')

age     sex     bmi  children smoker     region      charges  debt
1248   18  female  39.820         0     no  southeast   1633.96180   NaN
482    18  female  31.350         0     no  southeast   1622.18850   NaN
492    18  female  25.080         0     no  northeast   2196.47320   NaN
525    18  female  33.880         0     no  southeast  11482.63485   NaN
529    18    male  25.460         0     no  northeast   1708.00140   NaN
...   ...     ...     ...       ...    ...        ...          ...   ...
398    64    male  25.600         2     no  southwest  14988.43200   NaN
335    64    male  34.500         0     no  southwest  13822.80300   NaN
378    64  female  30.115         3     no  northwest  16455.70785   NaN
1265   64    male  23.760         0    yes  southeast  26926.51440   NaN
635    64    male  38.190         0     no  northeast  14410.93210   NaN

[1338 rows x 8 columns]

### 16. 평균구하기: mean

In [19]:
df.mean()

age            39.207025
bmi            30.663397
children        1.094918
charges     13270.422265
debt           13.000000
dtype: float64

### 17. 공분산 구하기: cov

In [20]:
df.cov()

age           bmi    children       charges         debt
age         197.401387      9.362337    0.719303  5.087480e+04     2.000000
bmi           9.362337     37.187884    0.093795  1.464730e+04     3.562500
children      0.719303      0.093795    1.453213  9.926742e+02     1.000000
charges   50874.802298  14647.304426  992.674197  1.466524e+08 -7789.029955
debt          2.000000      3.562500    1.000000 -7.789030e+03     1.000000

### 18. 남자, 여자가 각각 몇명있는지 구하기

In [21]:
pd.value_counts(df.sex)

male      676
female    662
Name: sex, dtype: int64

### 19. 자녀(children)이 3명인 데이터 출력하기

In [22]:
df[df['children'].isin([3])]

age     sex     bmi  children smoker     region      charges  debt
2      28    male  33.000         3     no  southeast   4449.46200   NaN
7      37  female  27.740         3     no  northwest   7281.50560   NaN
25     59  female  27.720         3     no  southeast  14001.13380   NaN
36     62  female  32.965         3     no  northwest  15612.19335   NaN
54     40  female  28.690         3     no  northwest   8059.67910   NaN
...   ...     ...     ...       ...    ...        ...          ...   ...
1301   62    male  30.875         3    yes  northwest  46718.16325   NaN
1314   30  female  23.655         3    yes  northwest  18765.87545   NaN
1320   31    male  31.065         3     no  northwest   5425.02335   NaN
1332   52  female  44.700         3     no  southwest  11411.68500   NaN
1333   50    male  30.970         3     no  northwest  10600.54830   NaN

[157 rows x 8 columns]

### 20. 자녀(children)가 5명 이상인 데이터 출력하기

In [23]:
df[df['children'] > 4]

age     sex     bmi  children smoker     region      charges  debt
32     19  female  28.600         5     no  southwest   4687.79700   NaN
71     31    male  28.500         5     no  northeast   6799.45800   NaN
166    20  female  37.000         5     no  southwest   4830.63000   NaN
413    25    male  23.900         5     no  southwest   5080.09600   NaN
425    45    male  24.310         5     no  southeast   9788.86590   NaN
438    52  female  46.750         5     no  southeast  12592.53450   NaN
568    49  female  31.900         5     no  southwest  11552.90400   NaN
640    33    male  42.400         5     no  southwest   6666.24300   NaN
877    33    male  33.440         5     no  southeast   6653.78860   NaN
932    46    male  25.800         5     no  southwest  10096.97000   NaN
937    39  female  24.225         5     no  northwest   8965.79575   NaN
969    39  female  34.320         5     no  southeast   8596.82780   NaN
984    20    male  30.115         5     no  northeast   4915.05985   NaN
1085   39  female  18.300         5    yes  southwest  19023.26000   NaN
1116   41    male  29.640         5     no  northeast   9222.40260   NaN
1130   39  female  23.870         5     no  southeast   8582.30230   NaN
1245   28    male  24.300         5     no  southwest   5615.36900   NaN
1272   43    male  25.520         5     no  southeast  14478.33015   NaN

### 21. bmi가  17 미만인 데이터 출력하기

In [24]:
df[df['bmi'] < 17]

age     sex     bmi  children smoker     region     charges  debt
172    18    male  15.960         0     no  northeast  1694.79640   NaN
428    21  female  16.815         1     no  northeast  3167.45585   NaN
1226   38    male  16.815         2     no  northeast  6640.54485   NaN

### 22. debt가 null이 아닌데이터 출력하기

In [25]:
df[df['debt'].notnull()]

age     sex     bmi  children smoker     region      charges  debt
3   33    male  22.705         0     no  northwest  21984.47061  12.0
5   31  female  25.740         0     no  southeast   3756.62160  13.0
8   37    male  29.830         2     no  northeast   6406.41070  14.0

### 23. 'age'가 60보다 큰 값만 출력하기

In [26]:
df[df['age'] > 60]

age     sex     bmi  children smoker     region      charges  debt
11     62  female  26.290         0    yes  southeast  27808.72510   NaN
26     63  female  23.085         0     no  northeast  14451.83515   NaN
33     63    male  28.310         0     no  northwest  13770.09790   NaN
36     62  female  32.965         3     no  northwest  15612.19335   NaN
62     64    male  24.700         1     no  northwest  30166.61817   NaN
...   ...     ...     ...       ...    ...        ...          ...   ...
1301   62    male  30.875         3    yes  northwest  46718.16325   NaN
1321   62    male  26.695         0    yes  northeast  28101.33305   NaN
1322   62    male  38.830         0     no  southeast  12981.34570   NaN
1325   61    male  33.535         0     no  northeast  13143.33665   NaN
1337   61  female  29.070         0    yes  northwest  29141.36030   NaN

[91 rows x 8 columns]